File: PDFNetC/PDFNetC64/Samples/JBIG2Test/CPP/JBIG2Test.cpp
//---------------------------------------------------------------------------------------
// Copyright (c) 2001-2020 by PDFTron Systems Inc. All Rights Reserved.
// Consult legal.txt regarding legal and license information.
//---------------------------------------------------------------------------------------

#include <iostream>
#include <PDF/PDFNet.h>
#include <PDF/PDFDoc.h>
#include <PDF/ElementBuilder.h>
#include <Filters/FilterReader.h>
#include <SDF/ObjSet.h>

using namespace std;
using namespace pdftron;
using namespace Filters;
using namespace PDF;
using namespace SDF;
using namespace Common;

// This sample project illustrates how to recompress bi-tonal images in an 
// existing PDF document using JBIG2 compression. The sample is not intended 
// to be a generic PDF optimization tool.
//
// You can download the entire document using the following link:
//   http://www.pdftron.com/net/samplecode/data/US061222892.pdf
//
int main(int argc, char *argv[]) 
{
	PDFNet::Initialize();
	
	try 
	{
		PDFDoc pdf_doc("../../TestFiles/US061222892-a.pdf");
		pdf_doc.InitSecurityHandler();

		SDFDoc& cos_doc = pdf_doc.GetSDFDoc();
		int num_objs = cos_doc.XRefSize();
		for(int i=1; i<num_objs; ++i) 
		{
			Obj obj = cos_doc.GetObj(i);
			if(obj && !obj.IsFree() && obj.IsStream()) 
			{
				// Process only images
				DictIterator itr = obj.Find("Subtype");
				if(!itr.HasNext() || strcmp(itr.Value().GetName(), "Image"))
					continue;
				
				Image input_image(obj);
				// Process only gray-scale images
				if(input_image.GetComponentNum() != 1)
					continue;
				int bpc = input_image.GetBitsPerComponent();
				if(bpc != 1)	// Recompress only 1 BPC images
					continue;

				// Skip images that are already compressed using JBIG2
				itr = obj.Find("Filter");
				if (itr.HasNext() && itr.Value().IsName() && 
					!strcmp(itr.Value().GetName(), "JBIG2Decode")) continue; 

				Filter filter=obj.GetDecodedStream();
				FilterReader reader(filter);


				ObjSet hint_set; 	// A hint to image encoder to use JBIG2 compression
				Obj hint=hint_set.CreateArray();
			
				hint.PushBackName("JBIG2");
				hint.PushBackName("Lossless");

				Image new_image = Image::Create(cos_doc, reader, 
					input_image.GetImageWidth(), 
					input_image.GetImageHeight(), 1, ColorSpace::CreateDeviceGray(), hint);

				Obj new_img_obj = new_image.GetSDFObj();
				itr = obj.Find("Decode");
				if(itr.HasNext())
					new_img_obj.Put("Decode", itr.Value());
				itr = obj.Find("ImageMask");
				if (itr.HasNext())
					new_img_obj.Put("ImageMask", itr.Value());
				itr = obj.Find("Mask");
				if (itr.HasNext())
					new_img_obj.Put("Mask", itr.Value());

				cos_doc.Swap(i, new_img_obj.GetObjNum());
			}
		}

		pdf_doc.Save("../../TestFiles/Output/US061222892_JBIG2.pdf", SDFDoc::e_remove_unused, 0);
	}
	catch(Common::Exception& e)
	{
		cout << e << endl;
		cout << "Please make sure that the pathname to the test file is correct." << endl;
	}
	catch(...)
	{
		cout << "Unknown Exception" << endl;
	}

	PDFNet::Terminate();
	return 0;
}

File: PDFNetC/PDFNetC64/Samples/JBIG2Test/JAVA/JBIG2Test.java
//---------------------------------------------------------------------------------------
// Copyright (c) 2001-2020 by PDFTron Systems Inc. All Rights Reserved.
// Consult legal.txt regarding legal and license information.
//---------------------------------------------------------------------------------------

import com.pdftron.filters.Filter;
import com.pdftron.filters.FilterReader;
import com.pdftron.pdf.*;
import com.pdftron.sdf.DictIterator;
import com.pdftron.sdf.Obj;
import com.pdftron.sdf.ObjSet;
import com.pdftron.sdf.SDFDoc;

// This sample project illustrates how to recompress bi-tonal images in an 
// existing PDF document using JBIG2 compression. The sample is not intended 
// to be a generic PDF optimization tool.
//
// You can download the entire document using the following link:
//   http://www.pdftron.com/net/samplecode/data/US061222892.pdf
public class JBIG2Test {

    public static void main(String[] args) {
        PDFNet.initialize();

        String input_path = "../../TestFiles/";
        String output_path = "../../TestFiles/Output/";

        try {
            PDFDoc pdf_doc = new PDFDoc(input_path + "US061222892-a.pdf");
            pdf_doc.initSecurityHandler();

            SDFDoc cos_doc = pdf_doc.getSDFDoc();
            int num_objs = (int) cos_doc.xRefSize();
            for (int i = 1; i < num_objs; ++i) {
                Obj obj = cos_doc.getObj(i);
                if (obj != null && !obj.isFree() && obj.isStream()) {
                    // Process only images
                    DictIterator itr = obj.find("Subtype");
                    if (!itr.hasNext() || !itr.value().getName().equals("Image"))
                        continue;

                    Image input_image = new Image(obj);
                    // Process only gray-scale images
                    if (input_image.getComponentNum() != 1)
                        continue;
                    int bpc = input_image.getBitsPerComponent();
                    if (bpc != 1)    // Recompress only 1 BPC images
                        continue;

                    // Skip images that are already compressed using JBIG2
                    itr = obj.find("Filter");
                    if (itr.hasNext() && itr.value().isName() &&
                            !itr.value().getName().equals("JBIG2Decode")) continue;

                    Filter filter = obj.getDecodedStream();
                    FilterReader reader = new FilterReader(filter);

                    ObjSet hint_set = new ObjSet();
                    Obj hint = hint_set.createArray(); // A hint to image encoder to use JBIG2 compression
                    hint.pushBackName("JBIG2");
                    hint.pushBackName("Lossless");

                    Image new_image = Image.create(cos_doc, reader,
                            input_image.getImageWidth(),
                            input_image.getImageHeight(), 1, ColorSpace.createDeviceGray(), hint);

                    Obj new_img_obj = new_image.getSDFObj();
                    itr = obj.find("Decode");
                    if (itr.hasNext())
                        new_img_obj.put("Decode", itr.value());
                    itr = obj.find("ImageMask");
                    if (itr.hasNext())
                        new_img_obj.put("ImageMask", itr.value());
                    itr = obj.find("Mask");
                    if (itr.hasNext())
                        new_img_obj.put("Mask", itr.value());

                    cos_doc.swap(i, new_img_obj.getObjNum());
                }
            }

            pdf_doc.save(output_path + "US061222892_JBIG2.pdf", SDFDoc.SaveMode.REMOVE_UNUSED, null);
            // output PDF pdf_doc
            pdf_doc.close();
        } catch (Exception e) {
            e.printStackTrace();
        }

        PDFNet.terminate();
    }

}
